File: 2015/Day3/main.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <set>
#include <utility>

int count_houses_with_presents(const std::string &directions)
{
    int x = 0, y = 0;
    std::set<std::pair<int, int>> visited;
    visited.insert({x, y});

    for (char direction : directions)
    {
        switch (direction)
        {
        case '^':
            y++;
            break;
        case 'v':
            y--;
            break;
        case '>':
            x++;
            break;
        case '<':
            x--;
            break;
        }
        visited.insert({x, y});
    }

    return visited.size();
}

int count_houses_with_presents2(const std::string &directions)
{
    int santa_x = 0, santa_y = 0, robot_x = 0, robot_y = 0;
    std::set<std::pair<int, int>> visited;
    visited.insert({0, 0});

    for (size_t i = 0; i < directions.length(); ++i)
    {
        int &x = (i % 2 == 0) ? santa_x : robot_x;
        int &y = (i % 2 == 0) ? santa_y : robot_y;

        switch (directions[i])
        {
        case '^':
            y++;
            break;
        case 'v':
            y--;
            break;
        case '>':
            x++;
            break;
        case '<':
            x--;
            break;
        }
        visited.insert({x, y});
    }

    return visited.size();
}

int main()
{
    std::ifstream file("input.txt");
    if (!file.is_open())
    {
        std::cerr << "Could not open the file!" << std::endl;
        return 1;
    }

    std::string directions;
    std::getline(file, directions);
    file.close();

    std::cout << "Part 1: " << count_houses_with_presents(directions) << std::endl;
    std::cout << "Part 2: " << count_houses_with_presents2(directions) << std::endl;

    return 0;
}

File: 2015/Day3/main.dart
import 'dart:collection';
import 'dart:io';

class Pair<T, U> {
  final T first;
  final U second;

  Pair(this.first, this.second);

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Pair &&
          runtimeType == other.runtimeType &&
          first == other.first &&
          second == other.second;

  @override
  int get hashCode => first.hashCode ^ second.hashCode;
}

int countHousesWithPresents(String directions) {
  int x = 0, y = 0;
  var visited = HashSet<Pair<int, int>>();
  visited.add(Pair(x, y));

  for (var direction in directions.split('')) {
    switch (direction) {
      case '^':
        y++;
        break;
      case 'v':
        y--;
        break;
      case '>':
        x++;
        break;
      case '<':
        x--;
        break;
    }
    visited.add(Pair(x, y));
  }

  return visited.length;
}

int countHousesWithPresents2(String directions) {
  int santaX = 0, santaY = 0, robotX = 0, robotY = 0;
  var visited = HashSet<Pair<int, int>>();
  visited.add(Pair(0, 0));

  for (int i = 0; i < directions.length; ++i) {
    int x, y;
    if (i % 2 == 0) {
      x = santaX;
      y = santaY;
    } else {
      x = robotX;
      y = robotY;
    }

    switch (directions[i]) {
      case '^':
        y++;
        break;
      case 'v':
        y--;
        break;
      case '>':
        x++;
        break;
      case '<':
        x--;
        break;
    }

    if (i % 2 == 0) {
      santaX = x;
      santaY = y;
    } else {
      robotX = x;
      robotY = y;
    }

    visited.add(Pair(x, y));
  }

  return visited.length;
}

void main() async {
  var file = File('input.txt');
  var directions = await file.readAsString();

  print('Part 1: ${countHousesWithPresents(directions)}');
  print('Part 2: ${countHousesWithPresents2(directions)}');
}
